src/py42/_internal/archive_access.py

```diff
@@ -125,61 +171,144 @@ def _get_children(self, node_id=None):
             show_deleted=True,
         )
 
-    @staticmethod
-    def _build_file_selection(file_path, file_type):
-        path_set = [{u"type": file_type, u"path": file_path, u"selected": True}]
-        # pass in dummy values; only used with progress indication, which we don't currently use
-        num_files = num_dirs = size = 1
-        return FileSelection(path_set, num_files, num_dirs, size)
+
+def _get_default_file_size():
+    return {u"numFiles": 1, u"numDirs": 1, u"size": 1}
 
 
-class RestoreJobManager(object):
+class _RestorePoller(object):
     JOB_POLLING_INTERVAL_SECONDS = 1
 
+    def __init__(self, storage_archive_client, device_guid, job_polling_interval=None):
+        self._storage_archive_client = storage_archive_client
+        self._device_guid = device_guid
+        self._job_polling_interval = (
+            job_polling_interval or self.JOB_POLLING_INTERVAL_SECONDS
+        )
+
+
+def _create_size_dict(job_id, size_response):
+    size_dict = json.loads(size_response.text)
+    size_dict[u"jobId"] = job_id
+    return size_dict
+
+
+def _print_file_size(size_dict):
+    debug.logger.debug(format_dict(size_dict))
+
+
+class FileSizePoller(_RestorePoller):
+    def __init__(
+        self, storage_archive_client, device_guid, job_polling_interval=None,
+    ):
+        super(FileSizePoller, self).__init__(
+            storage_archive_client, device_guid, job_polling_interval
+        )
+
+    def get_file_sizes(self, file_ids, timeout):
+        if not timeout:
+            # Skips file size calculation
+            return None
+
+        job_ids = self._start_poll(file_ids)
+        return self._wait_for_jobs(job_ids, timeout)
+
+    def _start_poll(self, file_ids):
+        job_ids = []
+        for file_id in file_ids:
+            response = self._storage_archive_client.create_file_size_job(
+                self._device_guid, file_id
+            )
+            job_id = response["jobId"]
+            job_ids.append(job_id)
+        return job_ids
+
+    def _wait_for_jobs(self, job_ids, timeout):
+        t0 = time.time()
+        sizes = []
+
+        # Waits until the job_ids stack is empty
+        while job_ids:
+            for job_id in job_ids:
+                response = self._get_job_status(job_id)
+                size_dict = _create_size_dict(job_id, response)
+                _print_file_size(size_dict)
+                if response[u"status"] == u"DONE":
+                    job_ids.remove(job_id)
+                    sizes.append(size_dict)
+
+            # File size calculation is taking too long.
+            if time.time() - t0 > timeout:
```

Comment:
Note that this is not a True timeout... Trying to use signal to achieve that but would need this to be on a separate process.  Maybe something to consider.  Could also check this in the `for` loop